### PR TITLE
Stampede update

### PR DESCRIPTION
**Pull request by marcomangano**
## Purpose
Updated Stampede 2 install and usage instructions. There are a couple of secondary TODOs left that can be addressed at a later stage.

## Type of change
- Documentation update

## Checklist

- [x] I have built the documentation locally

### CODE REVIEW

File: clusterSetup/stampedeSetup.rst [hunk 1]
```diff
@@ -11,6 +11,11 @@ This guide is intended to help users get setup with the MDO Lab software on the
 https://portal.xsede.org
 
 Once you have your XSEDE username setup, ask Prof. Martins to add you to the relevant allocation on XSEDE and confirm that you can log into the resource.
+Follow the email instructions to gain access to TACC system with your new alphanumerical username and set up the login authentication system (more info `here <https://portal.tacc.utexas.edu/tutorials/multifactor-authentication>`_).
```
Comment (nwu63) [inline clusterSetup/stampedeSetup.rst:14]: I think it's useful to explain the XSEDE SSO here as an alternative. I will add to this PR if you don't mind.

File: clusterSetup/stampedeSetup.rst [hunk 2]
```diff
@@ -24,28 +29,65 @@ Alternatively, you can directly ``ssh`` into stampede2 using
 
    ssh username@stampede2.tacc.utexas.edu
 
+Home, Work and Scratch directories
+----------------------------------
+
+Similarly to Great Lakes and other HPCs, you can access different file systems from your Stampede account. To learn more about the different directory, read `this section <https://portal.tacc.utexas.edu/user-guides/stampede2#overview-filesystems>`_ of the user guide.
+
+Use ``$HOME`` for ``pakages`` and ``repos`` folders.
```
Comment (nwu63) [inline clusterSetup/stampedeSetup.rst:37]: spelling: `packages`

File: clusterSetup/stampedeSetup.rst [hunk 3]
```diff
@@ -24,28 +29,65 @@ Alternatively, you can directly ``ssh`` into stampede2 using
...
+For our typical production jobs, with relatively limited memory and file I/O usage, it is fine to use the ``$WORK`` directory. It has a 1TB limit but it is never purged, while ``$SCRATCH`` has no memory limit but any file that is not executed or modified for more than 10 days will be deleted.
```
Comment (nwu63) [inline clusterSetup/stampedeSetup.rst:38]: "memory" usually refers to RAM, please change to "storage" which usually refers to HDD.

Should add warning about scratch. It's *usually* way more than 10 days, but AFAIK they do not let you know before deleting things, so do not store anything important there.

File: clusterSetup/stampedeSetup.rst [hunk 4]
```diff
@@ -24,28 +29,65 @@ Alternatively, you can directly ``ssh`` into stampede2 using
...
+
+Be careful to set ``-DCGNS_BUILD_CGNSTOOLS = 0`` to skip the installation of cgnstools.
+As you will use Intel compilers, you **must** set the compilers path by using these commands before configuring the build:
+
+.. code-block :: bash
+
+   export CC=$(which icc)
+   export FC=$(which ifort)
+
+If you forget to do so, the build will apparently succeed but you will not be able to use or import ``cgns`` with any of our tools, as the library will point to a default, non-existing compiler.
```
Comment (nwu63) [inline clusterSetup/stampedeSetup.rst:76]: This is not exactly correct. CGNS will by default compile using `/usr/bin/gcc`, which is the system default (not part of a module and always there regardless of what you load). These hard-coded paths are the problem. CGNS will usually compile fine, but since the rest of the code is compiled using a totally different compiler (and in this case Intel instead of GNU) there will be incompatibilities when compiling our codes against the previously compiled CGNS module. Setting these paths ensure that CGNS is compiled using the same compiler as the other codes.

File: clusterSetup/stampedeSetup.rst [hunk 5]
```diff
@@ -58,8 +100,10 @@ Environmental variables are placed in the if block under section 2:
    # Library Path for MPI (only needed if you are compiling PETSc)
    export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:$PETSC_DIR/$PETSC_ARCH/lib
```
Comment (nwu63) [inline clusterSetup/stampedeSetup.rst:101]: This section should be removed. We no longer recommend installing MPI with PETSc even if you are compiling your own PETSc on Stampede.

File: clusterSetup/stampedeSetup.rst [hunk 6]
```diff
@@ -58,8 +100,10 @@ Environmental variables are placed in the if block under section 2:
...
 
-   # Library Path for CGNSlib
-   export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:$HOME/packages/cgnslib_3.2.1/src
+   # -- CGNS
+   export CGNS_HOME=$HOME/packages/CGNS-3.3.0/opt-gfortran
+   export PATH=$PATH:$CGNS_HOME/bin
+   export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:$CGNS_HOME/lib
 
    # Path for repos directories
    export PYTHONPATH=$PYTHONPATH:$HOME/repos/
```
Comment (nwu63) [inline clusterSetup/stampedeSetup.rst:109]: This is no longer needed if you pip install, which is the currently-recommended approach.

File: clusterSetup/stampedeSetup.rst [hunk 7]
```diff
@@ -78,19 +122,29 @@ Lastly, the aliases are placed under section 3.
...
 
-Adjust directory names as needed. If you want to use the PETSc already compiled on stampede2, then you need to have ``module load petsc/3.7`` as mentioned above, and you no longer need the first three export statements.
+.. NOTE ::
+
+   ``TG-DDM140001`` refers to the MDO Lab allocation, it is not related to your specific user. You should not modify it unless you are accessing to another specific allocation.
 
 Running Jobs
 ------------
-Stampede2 uses Slurm rather than PBS (Moab or Torque). Also note that, it is generally advised to use SKX nodes rather than KNL for running MDOlab code, as they are more optimized for those architectures.
+Stampede2 uses Slurm rather than PBS (Moab or Torque). Also note that, it is generally advised to use SKX nodes rather than KNL for running MDO Lab code, as they are more optimized for those architectures.
```
Comment (nwu63) [inline clusterSetup/stampedeSetup.rst:137]: Now that Great Lakes is also using Slurm, we don't need the first sentence.

File: clusterSetup/stampedeSetup.rst (hunk 1 above)
Comment (marcomangano) [reply clusterSetup/stampedeSetup.rst:14]: Yes please, I will fix the other comments in the meantime

**Comment by marcomangano** [comment]
I agree we should discuss about the location of this guide. I don't think there is any reason to host the clusters set-up instructions on the website. I doubt this could be useful to anybody outside of the lab, in addition to your previous point. We can eventually make a sharable pdf version of the Stampede guide in case external collaborators intend to build MACH on it.

File: clusterSetup/stampedeSetup.rst (hunk 6 above)
Comment (nwu63) [reply clusterSetup/stampedeSetup.rst]: Sorry I wasn't very clear. I think we should leave it in, but commented out by default, and add a comment saying it's not necessary if all codes are installed using pip.

Comment (marcomangano) [reply clusterSetup/stampedeSetup.rst]: fixed!

File: clusterSetup/stampedeSetup.rst [hunk 8]
```diff
@@ -78,19 +122,29 @@ Lastly, the aliases are placed under section 3.
...
+Adjust directory names as needed. If you want to use the PETSc already compiled on stampede2, then you need to have ``module load petsc/3.11`` as mentioned above, and you no longer need the first three export statements.
+
+.. NOTE ::
+
+   ``TG-DDM140001`` refers to the MDO Lab allocation, it is not related to your specific user. You should not modify it unless you are accessing to another specific allocation.
 
 Running Jobs
 ------------
-Stampede2 uses Slurm rather than PBS (Moab or Torque). Also note that, it is generally advised to use SKX nodes rather than KNL for running MDOlab code, as they are more optimized for those architectures.
+Stampede2 uses Slurm as job scheduler. It is generally advised to use SKX nodes rather than KNL for running MDO Lab code, as they are more optimized for those architectures.
```
Comment (anilyil) [inline clusterSetup/stampedeSetup.rst:137]: Can you add a brief explanation of differences between SKX and KNL? SKX nodes have fewer cores per node, but each core runs at a higher clock speed. KNL has more cores per node, but they all run at a lower speed. Also, queue times on SKX is typically longer than KNL. Just this brief info, and a link to the actual architecture on the stampede user guide should be good enough (https://portal.tacc.utexas.edu/user-guides/stampede2#system-overview).

File: clusterSetup/stampedeSetup.rst [hunk 9]
```diff
@@ -78,19 +122,29 @@ Lastly, the aliases are placed under section 3.
    alias emn='emacs -nw'
    alias jstat='scontrol show job'
    alias iknl='idev -n 68 -N 1 -m 120 -A TG-DDM140001'
-   alias iskx='idev -p skx-dev -n 196 -N 4 -m 120 -A TG-DDM140001'
+   alias iskx='idev -p skx-dev -n 48 -N 1 -m 120 -A TG-DDM140001'
```
Comment (anilyil) [inline clusterSetup/stampedeSetup.rst:125]: if this is in public docs, we probably should not have the account number here. it most likely wont matter, but I would rather not include them

File: clusterSetup/stampedeSetup.rst [hunk 10]
```diff
@@ -78,19 +122,29 @@ Lastly, the aliases are placed under section 3.
...
    alias myqq='showq -u'
+   alias strtime='squeue --start -j'  # <jobID>, check estimated startime of your job
 
-Adjust directory names as needed. If you want to use the PETSc already compiled on stampede2, then you need to have ``module load petsc/3.7`` as mentioned above, and you no longer need the first three export statements.
+Adjust directory names as needed. If you want to use the PETSc already compiled on stampede2, then you need to have ``module load petsc/3.11`` as mentioned above, and you no longer need the first three export statements.
+
+.. NOTE ::
+
+   ``TG-DDM140001`` refers to the MDO Lab allocation, it is not related to your specific user. You should not modify it unless you are accessing to another specific allocation.
```
Comment (anilyil) [inline clusterSetup/stampedeSetup.rst:133]: again, I am for removing the account number

File: clusterSetup/stampedeSetup.rst [hunk 11]
```diff
@@ -78,19 +122,29 @@ Lastly, the aliases are placed under section 3.
...
 
 Running Jobs
 ------------
-Stampede2 uses Slurm rather than PBS (Moab or Torque). Also note that, it is generally advised to use SKX nodes rather than KNL for running MDOlab code, as they are more optimized for those architectures.
+Stampede2 uses Slurm as job scheduler. It is generally advised to use SKX nodes rather than KNL for running MDO Lab code, as they are more optimized for those architectures.
+
+.. NOTE ::
+
+   Your jobs will most likely stay in the queue from several hours up to a day, depending on the resources you are requesting. We do not have specific tips in this sense, except for what already reported in the User guide. Don't ask for more resources than you actually need! You can get a glimpse of Stampede 2 current usage on this `system monitor <https://portal.tacc.utexas.edu/system-monitor>`_.
```
Comment (anilyil) [inline clusterSetup/stampedeSetup.rst:141]: can you include a short note on the prioritization system? When you use this system less, your priority goes up, so your jobs start quickly. If you try to use it too much in a burst, then priority goes down a lot so you wait unnecessarily long queues. The best approach is to keep your utilization of the system as constant as possible. This is not always possible, but the users should at least know about the prioritization system. I think this prioritization is made on the whole account level (jrram's allocation), vs on a per-user basis, but dont quote me on this because I may be wrong.

File: clusterSetup/stampedeSetup.rst [hunk 12]
```diff
@@ -108,4 +162,8 @@ Example run script:
 
     # Launch MPI code...
 
-    ibrun python myscript.py   # ibrun is used instead of mpirun/mpiexec on stampede
+    ibrun -n 240 python myscript.py   # ibrun is used instead of mpirun/mpiexec on stampede
```
Comment (anilyil) [inline clusterSetup/stampedeSetup.rst:165]: why `-n 240` after `ibrun`?

File: clusterSetup/stampedeSetup.rst [hunk 13]
```diff
@@ -108,4 +162,8 @@ Example run script:
...
+
+.. TIP ::
+
+   Interactive jobs are a useful resource. There is only a time limit (120 minutes) and you can request a high number of nodes. The queue time varies from few seconds to few minutes. Although it is not recommended to use these jobs for production (unless, for example, you have to run a set of quick ADflow runs), it is **strongly** recommended to test your run scripts here before you submit a regular job. You don't want to wait a day for your job to start and then have it crashing after a few seconds for some trivial coding mistake.
```
Comment (anilyil) [inline clusterSetup/stampedeSetup.rst:169]: this is not accurate, and seems like a common misconception users have. I am pretty sure this holds for every cluster we use: An interactive job does not mean its a development job. Any job can be either interactive, i.e. you have a terminal and you can run commands, when you exit the terminal, the job finishes. It's like having an ssh connection. Or it can be a batch job (not sure if this is the correct name), but in this case, whatever is in your job script is run. Don't remember which one, but some clusters let you ssh into the nodes that are running your jobs so you can monitor the processes. 

Besides this, there is the differentiation of normal vs development queues. These simply provide the limits on resources you can use, and they have different priorities. 

Bottom line: A normal job can be run interactively, and a development job can be started with a job script.

File: clusterSetup/stampedeSetup.rst (hunk 9 above)
Comment (anilyil) [reply clusterSetup/stampedeSetup.rst:125]: I also have aliases called iskx1, iskx2 etc., and the number means how many nodes I request for that. Can be useful for some cases.

File: clusterSetup/stampedeSetup.rst [hunk 14]
```diff
@@ -12,40 +12,85 @@ https://portal.xsede.org
...
 
 Setting up MDO Lab software
 ---------------------------
 For most cluster setups, you need to follow the basic steps outlined in :ref:`installFromScratch`.
 
-It is advised to install all the MDOlab code under ``$HOME/repos``. A few differences to note:
+Again, it is advised to install all the MDO Lab code under ``$HOME/repos``. A few differences to note:
 
-- :ref:`PETSc/MPI <install_petsc>` has already been compiled, so it's possible to omit them during the installation process, and load the required modules with, say, ``module load petsc/3.7``
+- :ref:`PETSc/MPI <install_petsc>` has already been compiled, so it's possible to omit them during the installation process, and load the required modules with, say, ``module load petsc/3.11``
```
Comment (anilyil) [inline clusterSetup/stampedeSetup.rst:55]: I would change the language here. The default should be to use the system MPI. These installations are often provided the vendor who built the cluster, and are optimized for each cluster. I would remove the word "possible to omit", and say "use the system MPI". Besides, you can build your own petsc, but it's probably better to use whats provided in the system. So in my opinion, system MPI is a must, system petsc is optional.

File: clusterSetup/stampedeSetup.rst [hunk 15]
```diff
@@ -12,40 +12,85 @@ https://portal.xsede.org
...
+.. WARNING ::
+
+   ``$SCRATCH`` purge might occur after more than 10 days but you will not be warned in advance, so be extra careful.
+
+.. TIP ::
+
+   It is good practice to always download or backup your output files as soon as the job is completed.
+
+.. TODO : add file backup tips
+.. TODO : using transfer nodes
```
Comment (anilyil) [inline clusterSetup/stampedeSetup.rst:47]: just add links on how to use globus (https://portal.tacc.utexas.edu/user-guides/stampede2#transferring-files-using-globus) and a short tutorial of psftp (check this one: https://www.ssh.com/ssh/putty/putty-manuals/0.68/Chapter6.html I am not sure but this may be just for putty, which most people here dont use. Still, information can be useful). We can refine this later but just having the links should be good enough for now.

File: clusterSetup/stampedeSetup.rst [hunk 16]
```diff
@@ -12,40 +12,85 @@ https://portal.xsede.org
...
 .. code-block:: bash
 
-   module load git
-   module load intel/17.0.4
-   module load python/2.7.13
-   module load petsc/3.7              # If you want to use pre-compiled PETSc
+   module load git/2.24.1
+   module load intel/18.0.2
+   module load petsc/3.11              # If you want to use pre-compiled PETSc
+
```
Comment (anilyil) [inline clusterSetup/stampedeSetup.rst:90]: can you list the system MPI we should import here? Is it imported by default?

File: clusterSetup/stampedeSetup.rst (hunk 9 above)
Comment (nwu63) [reply clusterSetup/stampedeSetup.rst:125]: I have a script (for great lakes currently) which will request interactive jobs based on the following inputs:
- queue
- time
- account
- nprocs

Something like that for stampede would be useful also.

File: clusterSetup/stampedeSetup.rst (hunk 13 above)
Comment (nwu63) [reply clusterSetup/stampedeSetup.rst:169]: All good, however I think what you said is really applicable to only Great Lakes. On Stampede the queue time for the normal queue is to long to submit an interactive job IMO, but maybe you can do it. So for these clusters, it's still reasonable to say you use interactive job on dev queue and batch job on normal queue.

Also, all HPC should provide the ability to ssh into your compute nodes. You just need to figure out the hostname, and whichever command that lists your running jobs should do that. The system will also prevent you from being able to ssh into nodes that you don't have jobs running on.

File: clusterSetup/stampedeSetup.rst (hunk 9 above)
Comment (marcomangano) [reply clusterSetup/stampedeSetup.rst:125]: please add the scripts/aliases here in the comments, I will include them later.

File: clusterSetup/stampedeSetup.rst (hunk 12 above)
Comment (marcomangano) [reply clusterSetup/stampedeSetup.rst:165]: Yeah maybe I can remove it. It's basically the same mpirun option

File: clusterSetup/stampedeSetup.rst (hunk 16 above)
Comment (marcomangano) [reply clusterSetup/stampedeSetup.rst:90]: the default is `impi/18.0.2`

Comment (nwu63) [reply clusterSetup/stampedeSetup.rst:90]: We should include that since we need to specify the version which is compatible with intel/18.0.2.

File: clusterSetup/stampedeSetup.rst (hunk 13 above)
Comment (marcomangano) [reply clusterSetup/stampedeSetup.rst:169]: So, in short, how would you like to rephrase this? Being honest I was not fully aware about the normal vs dev queuing, my message was more "run a quick interactive job before you batch something". My main point is that interactive (or I should say development?) jobs have a time limit so, unless you can run a bunch of analyses or a quick optimization run, you still should "batch" your regular jobs.

Comment (nwu63) [reply clusterSetup/stampedeSetup.rst:169]: I think the explanation should be done in the HPC guide, and not here. We should just explain "here are the queues" and "here is how to submit interactive job and here's how to submit batch job", and just link to different sections of the user guide.

So, maybe just add a few links to Stampede-specific info, and leave the rest of the explanation for the HPC guide?

Comment (marcomangano) [reply clusterSetup/stampedeSetup.rst:169]: That is good with me

Comment (nwu63) [reply clusterSetup/stampedeSetup.rst:169]: See for example [this](https://github.com/mdolab/docs/blob/master/clusterSetup/GreatlakesSetup.rst#partitions) in the GL page. We could just have a table on hardware, a table on queues, and a note/section on interactive jobs.

Comment (marcomangano) [reply clusterSetup/stampedeSetup.rst:169]: I am commenting out the section for now, we can use it on the HPC guide at a later stage.
